Address final review of #2638

cprecioso opened the pull request:
Review changes for #2638

Martinsos reviewed (commented):
Left some comments @cprecioso , I like where we are going, I think we are quite closer to a finished thing!

Except for what I commented on, one main thing that I am suspicious a bit of is the current `LanguageInfo` situation. That whole part is certainly better now, as you said it is more explicit, there are less hidden assumptions. But on the other hand, this centralization of LanguageInfo seems a bit weird to me, a bit artifical and too emphasized. I tryed playing a bit with replacing that structure with functions like `function tsToJsExt (tsExt: "ts" | "tsx"): "js" | "jsx"` and `function tsExtIsJsx (tsExt: "ts" | "tsx"): boolean` but I am not sure that is the right direction either.
Maybe that is ok all together -> I would say, let's take care of everyting that is outstanding, then at the very end we can take one last look at the LanguageInfo situation and see if we want to do something there or not, it will be clearer situation then.

cprecioso commented on web/src/remark/auto-js-code.ts:
```diff
@@ -35,70 +34,107 @@ Output:
 
 */
 
+// Type-only empty import to register MDX node types into mdast.
```
Not needed in latest version, removed it

cprecioso commented on web/src/remark/auto-js-code.ts:
```diff
@@ -35,70 +34,107 @@ Output:
+  tsx: { jsx: true, outputLang: 'jsx' },
+} satisfies Record<string, SupportedLanguageInfo>
+
+const SUPPORTED_EXTENSIONS = {
```
If we're being explicit and extensible, I'd rather keep it like this. For example `.ctsx` or `.mtsx` are valid TypeScript extension, and if we'd be adding this programmatically it'd end up like

```ts
const makeAllowedExtensionsForLang = (langName: string) =>
  ["c","m",""].flatMap(prefix =>
    ["x", ""].map(suffix =>
      `.${prefix}${langName}${suffix}`
```

I don't think this is very understandable (or grep-able!), and not much better than the `.replace("t", "j")` thing we had before.

Martinsos replied:
As I mentioned in another comment, I still have this hunch that we can do a bit better here, but might be wrong and I think it is not worth pursuing it, so all good with me on this one!

RAW

cprecioso commented on web/src/remark/code-with-hole.ts:
```diff
@@ -25,24 +26,32 @@ Output:
+// Wrapped in \b to denote a word boundary.
 const META_FLAG_REGEX = /\bwith-hole\b/
-const HOLE_IDENTIFIER_REGEX = /\bhole\b/
+const HOLE_IDENTIFIER_REGEX = /\b\$HOLE$\b/
```
I forgot to escape the last `$`

```diff
-const HOLE_IDENTIFIER_REGEX = /\b\$HOLE$\b/
+const HOLE_IDENTIFIER_REGEX = /\b\$HOLE\$\b/
```

cprecioso commented on web/src/remark/util/code-blocks.ts:
```diff
@@ -1,36 +1,34 @@
-  file: VFile,
-  supportedLanguages: Set<T>
-): asserts node is md.Code & { lang: T } {
+export function assertSupportedLanguage<
```
I don't agree. Mainly because I do think that the most informative error message here is `Unsupported language`, so it must be checking if it is supported. (otherwise which? "Unknown"? Doesn't feel right)

> What does it mean it is supported? By whom? 

The caller function decides that. That's why it passes the `supportedLanguages` array.

This is supposed to be a util function for plugins, not a standalone one. I think it's valid that it has terms from the plugin domain.

Martinsos replied:
I get what you mean! And yeah it is certainly ok for it to use terms in the context of the plugin domain, I agree.

That said I am still not super convinced on having "supported" here. As you said yourself, this meaning is coming from the caller. If so, why not keep that meaning in the "domain" of the caller, to put it that way?

Actually, why "supported" at all? Does that mean "supported by this plugin"? OK, if so that makes sense -> wasn't immediate connection for me, but it makes sense, I agree.

But I still think you can drop this word from here, make function name clearer, and not loose anything regarding the error messaging clarity (which I agree is a strong argument, as you made it above):
- name function `assertCodeBlockIsInLanguage`. Name is super clear, certainly not worse than current name, and I would say it also skips the somewhat ambiguous term "supported language" (I get it though if you think it is not ambiguous in this context, that is valid argument).
- Change the error message to "Expected code block to be in one of languages ..., but got ...". I don't think you lose anything by not having word "supported" here. If it is really taht important to have word "supported", becaues in the caller that words has important meaning, than the caller can easily add more to the error message by catching and throwing, and it can do it with more meaning than this function ever could.

That said, this is really quite a nitpick and I am not super convinced the change I suggested is really important, so I am completely fine with you leaving it as is. I wrote the above more for the sake of argumentation/discussion, not because I think it is that important. I wrote it all already so I will leave it :D. Word "supported" indeed does have some meaning here in this plugin context and it can work just fine.

RAW

cprecioso commented:
@Martinsos addressed the comments, please review

---

> I tryed playing a bit with replacing that structure with functions like function tsToJsExt (tsExt: "ts" | "tsx"): "js" | "jsx" and function tsExtIsJsx (tsExt: "ts" | "tsx"): boolean but I am not sure that is the right direction either.

I think Haskell-isms are leading you on here, I wouldn't say this kind of solution is idiomatic JS/TS at all.

> seems a bit weird to me, a bit artifical and too emphasized

Well, yes 😅 If we want to operate at this level of abstraction, we need a more-or-less central way of defining the transformation. And right now, it does look like overkill for just two possible alternatives. If we want to make it simpler I think I'd end up slipping back to something very similar to what we had.

Martinsos commented on web/src/remark/util/code-blocks.ts:
```diff
@@ -1,36 +1,51 @@
   }
 }
+
+function throwLangError(
```
Curious: why not put this function inside the assertSupportedLanguage function? I guess it might have a bit of performance impact, but I imagine that is negligible, while you potentially get nicer scoping, in the sense that it is clear it is helper function inside of this bigger function and is not intended to be used anywhere else.

This one can though probably stand on its own, so all fine with me if you like this better, but I would be tempted to put it inside, so just curious to hear your reasoning for the opposite.

cprecioso replied:
Personal preference, vaguely based on perf. I am generally averse to creating utility functions inside other functions (esp. in the hot path). This used to make sense due to the function having to be reconstructed every time, but probably JS engines are intelligent enough now to work around it.

No strong opinion on my side, but if there's also no strong opinion from your side I'll leave it as-is.

Martinsos commented:
> @Martinsos addressed the comments, please review
> 
> > I tryed playing a bit with replacing that structure with functions like function tsToJsExt (tsExt: "ts" | "tsx"): "js" | "jsx" and function tsExtIsJsx (tsExt: "ts" | "tsx"): boolean but I am not sure that is the right direction either.
> 
> I think Haskell-isms are leading you on here, I wouldn't say this kind of solution is idiomatic JS/TS at all.
> 
> > seems a bit weird to me, a bit artifical and too emphasized
> 
> Well, yes 😅 If we want to operate at this level of abstraction, we need a more-or-less central way of defining the transformation. And right now, it does look like overkill for just two possible alternatives. If we want to make it simpler I think I'd end up slipping back to something very similar to what we had.

Yes it is probably a bit of Haskellisms as you said :D. I do find Haskell often to influence my code in other languages for better though, so I don't disregard them but usually follow them, to see where it brings me.
In this case, I agree that didn't feel like the right direction at the end, as I noted above. And I completely get what you did with the current approach and think it is a nice improvement.
Still, I have something gnawing at me that there is somewhat better solution just in reach, feels like there is too much complexity. But I do have to admit I haven't managed to see it quickly, and it might be a fake hunch, so I am happy with the current solution -> it is explicit and clear and that is most important, and since it is relatively local and scoped, a bit of potentially redundant complexity / non-cohesiveness can't do much damage. And exploring something better (if there is anything at all) would probably take time that is just not worth it.

Ok, good with me, nice stuff!

Martinsos commented on web/src/remark/util/code-blocks.ts:
```diff
@@ -1,36 +1,51 @@
+
   return (node: md.Nodes): node is md.Code & { meta: string } =>
-    node.type === "code" && node.meta && metaFlag.test(node.meta);
+    node.type === "code" && node.meta && metaFlagRegexp.test(node.meta);
```
If I am correct, this will also match on `title="auto-js.tsx"`, even though we don't want it, right?

I think we discussed it previously (can't find the comment now) and that you said it is not worth putting more effort into it. But that means we are leaving the code here that we know works incorrectly in some cases. I would say, either we document it here, so at least people in the future know why we did so, or even better, we fix it now. Is it that much work really? We just need to split the meta by `fieldwithvalue="value"` or `fieldwithnovalue` and then check those. Shouldn't be that hard, GPT can probably spit out the function to do it in one shot, and just regexes should be enough. If we can do that, I would vote for doing it now, so we can rest in piece knowing we left code that will always work correctly.

Martinsos reviewed (commented):
@cprecioso ok we are very close now! I think there is nothing major left,expect for one thing that I would really want us to fix before we close this, and that is making that function for chekcing the meta flag work correctly and not give false positives.

cprecioso commented on web/src/remark/util/code-blocks.ts:
```diff
@@ -1,36 +1,66 @@
+
+const CODE_BLOCK_META_REGEX = /([^\s=]+)(?:="([^"]*)")?/g;
+export function parseCodeBlockMetaString(
+  metaString: string | undefined,
```
The intention was to just pass `codeNode.meta` without having to do more checks. I think there's a case to be made that this could accept the `undefined`, but tbh I don't really mind anyway.

cprecioso commented on web/src/remark/util/code-blocks.ts:
```diff
@@ -1,36 +1,66 @@
+  return result;
+}
+
+export function stringifyCodeBlockMetaString(
```
I used the `parse`/`stringify` convention from `JSON.parse()`/`JSON.stringify()` and the `unified` library ecosystem. It is just next to the `parse` function so it is clear that they are the inverse of each other.

> what does it mean to stringify a (code block meta) string? Insn't it already a string?

The name is slightly wrong, it is not a string already but a `Map`. I fixed it like this:
```diff
-stringifyCodeBlockMetaString
+stringifyCodeBlockMeta
```

cprecioso commented:
@Martinsos ready

cprecioso commented:
@Martinsos check it out
